crypto/polyval-generic.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * POLYVAL: hash function for GCM-SIV.
+ * Code based on crypto/ghash-generic.c
+ *
+ * Copyright (c) 2007 Nokia Siemens Networks - Mikko Herranen <mh1@iki.fi>
+ * Copyright (c) 2009 Intel Corp.
+ *   Author: Huang Ying <ying.huang@intel.com>
+ * Copyright (c) 2021 Google LLC
+ */
+
+#include <crypto/algapi.h>
+#include <crypto/gf128mul.h>
+#include <crypto/polyval.h>
+#include <crypto/internal/hash.h>
+#include <linux/crypto.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+
+static int polyval_init(struct shash_desc *desc)
+{
+	struct polyval_desc_ctx *dctx = shash_desc_ctx(desc);
+
+	memset(dctx, 0, sizeof(*dctx));
+
+	return 0;
+}
+
+static void reverse_bytes(u8 *buf) {
```

Comment:
This code and `crypto_xor_reverse` could be simplified into a single function that efficiently reverses an arbitrary length bytestring.

Reply:
Why does it need to be arbitary length?  It looks like in both cases, only `POLYVAL_BLOCK_SIZE` bytes are reversed.

I think something like the following that's specialized for 16-byte blocks would be better suited for this:

```
static void reverse_block(u8 block[POLYVAL_BLOCK_SIZE])                          
{                                                                                
        u64 *p1 = (u64 *)block;                                                  
        u64 *p2 = (u64 *)&block[8];                                              
        u64 a = get_unaligned(p1);                                               
        u64 b = get_unaligned(p2);                                               
                                                                                 
        put_unaligned(swab64(a), p2);                                            
        put_unaligned(swab64(b), p1);                                            
}                                                                                
```